onnx/defs/generator/defs.cc

```diff
@@ -225,3 +225,56 @@ TensorProto message, and be valid as an output type.
         }
         propagateShapeFromInputToOutput(ctx, 0, 0);
     });
+
+ONNX_OPERATOR_SCHEMA(Multinomial)
+    .SetDoc(R"DOC(
+Generate a tensor of samples from a multinomial distribution according to the probabilities
+of each of the possible outcomes.
+)DOC")
+    .Attr(
+        "sample_size",
+        "Number of times to sample.",
+        AttributeProto::INT,
+        static_cast<int64_t>(0))
+    .Attr(
+        "seed",
+        "(Optional) Seed to the random generator, if not specified we will auto generate one.",
+        AttributeProto::FLOAT,
+        OPTIONAL)
+    .Input(
+        0,
+        "input",
+        "Input tensor with shape [batch_size, class_size], where class_size is the number of all possible outcomes.",
+        "T1")
+    .Output(
+        0,
+        "output",
+        "Output tensor with shape [batch_size, sample_size], where sample_size is the number of times to sample. ",
+        "T2")
+    .TypeConstraint(
+        "T1",
+        {"tensor(float16)", "tensor(float)", "tensor(double)"},
+        "Constrain input types to float tensors.")
+    .TypeConstraint(
+        "T2",
+        {"tensor(int32)", "tensor(int64)"},
+        "Constrain output types to integral tensors.")
+    .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+      auto output_type = ctx.getOutputType(0);
+      auto input_type = ctx.getInputType(0);
+      if (TypeProto::kTensorType != output_type->value_case() ||
+          TypeProto::kTensorType != input_type->value_case()) {
+        return;
+      }
+      auto sample_size = ctx.getAttribute("sample_size");
+      if (nullptr == sample_size || // attribute not present
```

Comment:
This can't be true for a valid Operator. I suggest shape inference code don't do validation and always operate on the assumption that the model file has passed the checker. (At least here don't silently return, it should raise some errors or similar) @anderspapitto @gramalingam What's the error reporting mechanism in shape inference?

Reply:
About error handling in shape inference, I'd suggest to add "Status" as return value. @anderspapitto  and @gramalingam  let me know whether you're OK on it or not please. The "Status" class was added recently. I had ever given @anderspapitto this proposal (using status as return value) when the interface was checked in.

Reply:
I think returning a Status code is reasonable. While I agree that it is reasonable to assume that the model file has passed the checker before inference is invoked, it is still a good idea to safeguard the inference code against null-dereference etc. (E.g., this ensures that even if the op signature is changed to make an attribute optional, we don't fail with a null dereference.) Shape inference is (currently) viewed as optional (for optimization): there will be models where shape inference won't successfully infer shapes. For this reason, returning silently is a reasonable compromise for shape inference (in such edge cases). At least, that's my personal opinion.

Reply:
You can use getAttribute(ctx, "sample_size", -1) and return if the value is negative.

Reply:
@bddppq do we have a list or documentation somewhere of what guarantees the checker provides about the model?

Reply:
@anderspapitto Nope, but it's a good idea to document this.